File: OJ practices/LeetCode题目集/671. Second Minimum Node In a Binary Tree(easy).cpp
/*
 * @Autor: SourDumplings
 * @Date: 2019-10-13 20:15:38
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/second-minimum-node-in-a-binary-tree/
 */

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    int findSecondMinimumValue(TreeNode *root)
    {
        if (!root || !root->left)
        {
            return -1;
        }
        int m = root->val;
        int l = findSecond(root->left, m);
        int r = findSecond(root->right, m);
        return l < r && l != -1 || r == -1 ? l : r;
    }

    int findSecond(TreeNode *node, int limit)
    {
        int v = node->val;
        if (v > limit)
        {
            return v;
        }
        else if (node->left)
        {
            int l = findSecond(node->left, limit);
            int r = findSecond(node->right, limit);
            return l < r && l != -1 || r == -1 ? l : r;
        }
        return -1;
    }
};

File: OJ practices/LeetCode题目集/671. Second Minimum Node In a Binary Tree(easy).java
/*
 * @Author: SourDumplings
 * @Date: 2019-10-13 20:38:21
 * @Link: https://github.com/SourDumplings/
 * @Email: changzheng300@foxmail.com
 * @Description: https://leetcode.com/problems/second-minimum-node-in-a-binary-tree/
 */

/**
* Definition for a binary tree node.
* public class TreeNode {
*     int val;
*     TreeNode left;
*     TreeNode right;
*     TreeNode(int x) { val = x; }
* }
*/
class Solution
{
    public int findSecondMinimumValue(TreeNode root)
    {
        if (root == null || root.left == null)
        {
            return -1;
        }
        int m = root.val;
        int l = findSecond(root.left, m);
        int r = findSecond(root.right, m);
        return l < r && l != -1 || r == -1 ? l : r;
    }

    private int findSecond(TreeNode node, int m)
    {
        int v = node.val;
        if (v > m)
        {
            return v;
        }
        else if (node.left != null)
        {
            int l = findSecond(node.left, m);
            int r = findSecond(node.right, m);
            return l < r && l != -1 || r == -1 ? l : r;
        }
        return -1;
    }
}